File: script/get_token.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");
const crypto = require("crypto");
const { exec } = require("child_process");

const AUTH_PAGE_URL_TEMPLATE = "https://app.mspbots.ai/datacli-auth?code={code}";
const TOKEN_URL_TEMPLATE = "https://owlstg.mspbots.ai//owl-agent/api/v1/auth_token/{code}";
const MAX_RETRIES = 60;
const POLL_INTERVAL_MS = 1000;
const CONFIG_PATH = path.resolve(__dirname, "..", "config", "config.yml");

function generateCode() {
  return crypto.randomUUID().replace(/-/g, "");
}

function openAuthPage(code) {
  const authUrl = AUTH_PAGE_URL_TEMPLATE.replace("{code}", code);
  const escapedUrl = `"${authUrl.replace(/"/g, '\\"')}"`;

  let cmd;
  if (process.platform === "win32") {
    cmd = `start "" ${escapedUrl}`;
  } else if (process.platform === "darwin") {
    cmd = `open ${escapedUrl}`;
  } else {
    cmd = `xdg-open ${escapedUrl}`;
  }

  exec(cmd, (error) => {
    if (error) {
      console.warn(`Failed to open browser automatically: ${error.message}`);
      console.warn(`Please open this URL manually: ${authUrl}`);
    }
  });

  return authUrl;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function pollAuthToken(code, maxRetries = MAX_RETRIES, intervalMs = POLL_INTERVAL_MS) {
  const tokenUrl = TOKEN_URL_TEMPLATE.replace("{code}", code);

  for (let attempt = 1; attempt <= maxRetries; attempt += 1) {
    try {
      const response = await fetch(tokenUrl, { method: "GET" });
      if (response.status === 200) {
        return await response.text();
      }
    } catch (_) {
      // Ignore transient network errors and retry.
    }

    if (attempt < maxRetries) {
      await sleep(intervalMs);
    }
  }

  throw new Error(`Polling failed: no successful response (200) within ${maxRetries} attempts.`);
}

function yamlScalar(value) {
  if (value === null || value === undefined) {
    return "null";
  }
  if (typeof value === "boolean") {
    return value ? "true" : "false";
  }
  if (typeof value === "number") {
    return String(value);
  }
  return JSON.stringify(String(value));
}

function toYamlLines(data, indent = 0) {
  const lines = [];
  for (const [key, value] of Object.entries(data)) {
    const prefix = " ".repeat(indent);
    if (value && typeof value === "object" && !Array.isArray(value)) {
      lines.push(`${prefix}${key}:`);
      lines.push(...toYamlLines(value, indent + 2));
    } else {
      lines.push(`${prefix}${key}: ${yamlScalar(value)}`);
    }
  }
  return lines;
}

function updateConfig(token, userInfo) {
  const configData = {
    mspbots: {
      token,
      user_info: userInfo,
    },
  };
  const yamlText = `${toYamlLines(configData).join("\n")}\n`;
  fs.writeFileSync(CONFIG_PATH, yamlText, "utf8");
}

async function main() {
  const code = generateCode();
  const authUrl = openAuthPage(code);
  console.log(`Browser opened. Please complete authorization: ${authUrl}`);

  try {
    const result = await pollAuthToken(code);
    let payload;
    try {
      payload = JSON.parse(result);
      console.log(JSON.stringify(payload, null, 2));
    } catch (_) {
      console.log(result);
      throw new Error("Invalid response: expected a JSON payload.");
    }

    if (!payload.ok) {
      throw new Error(`Authorization failed: ${payload.message || "unknown error"}`);
    }

    const token = payload.token;
    const userInfo = payload.user_info;
    if (!token || !userInfo || typeof userInfo !== "object" || Array.isArray(userInfo)) {
      throw new Error("Invalid response: missing token or user_info.");
    }

    updateConfig(token, userInfo);
    console.log(`Config updated: ${CONFIG_PATH}`);
  } catch (err) {
    console.error(err.message);
    process.exit(1);
  }
}

main();

File: script/get_token.py
import json
from pathlib import Path
import time
import uuid
import webbrowser
from urllib import error, request


AUTH_PAGE_URL_TEMPLATE = "https://app.mspbots.ai/datacli-auth?code={code}"
TOKEN_URL_TEMPLATE = "https://owlstg.mspbots.ai//owl-agent/api/v1/auth_token/{code}"
MAX_RETRIES = 60
POLL_INTERVAL_SECONDS = 1
CONFIG_PATH = Path(__file__).resolve().parents[1] / "config" / "config.yml"


def generate_code() -> str:
    return uuid.uuid4().hex


def open_auth_page(code: str) -> str:
    auth_url = AUTH_PAGE_URL_TEMPLATE.format(code=code)
    webbrowser.open(auth_url)
    return auth_url


def poll_auth_token(code: str, max_retries: int = MAX_RETRIES, interval: int = POLL_INTERVAL_SECONDS):
    token_url = TOKEN_URL_TEMPLATE.format(code=code)

    for attempt in range(1, max_retries + 1):
        try:
            with request.urlopen(token_url, timeout=10) as resp:
                body = resp.read().decode("utf-8", errors="ignore")
                if resp.status == 200:
                    return body
        except error.HTTPError as exc:
            if exc.code == 200:
                return exc.read().decode("utf-8", errors="ignore")
        except error.URLError:
            pass

        if attempt < max_retries:
            time.sleep(interval)

    raise RuntimeError(f"Polling failed: no successful response (200) within {max_retries} attempts.")


def _yaml_scalar(value):
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return str(value)
    return json.dumps(str(value), ensure_ascii=False)


def _to_yaml_lines(data, indent=0):
    lines = []
    for key, value in data.items():
        prefix = " " * indent
        if isinstance(value, dict):
            lines.append(f"{prefix}{key}:")
            lines.extend(_to_yaml_lines(value, indent + 2))
        else:
            lines.append(f"{prefix}{key}: {_yaml_scalar(value)}")
    return lines


def update_config(token: str, user_info: dict):
    config_data = {
        "mspbots": {
            "token": token,
            "user_info": user_info,
        }
    }
    yaml_text = "\n".join(_to_yaml_lines(config_data)) + "\n"
    CONFIG_PATH.write_text(yaml_text, encoding="utf-8")


def main():
    code = generate_code()
    auth_url = open_auth_page(code)
    print(f"Browser opened. Please complete authorization: {auth_url}")

    try:
        result = poll_auth_token(code)
        try:
            payload = json.loads(result)
            print(json.dumps(payload, ensure_ascii=False, indent=2))
        except json.JSONDecodeError:
            print(result)
            raise RuntimeError("Invalid response: expected a JSON payload.")

        if not payload.get("ok"):
            raise RuntimeError(f"Authorization failed: {payload.get('message', 'unknown error')}")

        token = payload.get("token")
        user_info = payload.get("user_info")
        if not token or not isinstance(user_info, dict):
            raise RuntimeError("Invalid response: missing token or user_info.")

        update_config(token, user_info)
        print(f"Config updated: {CONFIG_PATH}")
    except RuntimeError as exc:
        print(str(exc))
        raise SystemExit(1) from exc


if __name__ == "__main__":
    main()
